python/gstaichi/lang/_wrap_inspect.py

```diff
@@ -186,4 +188,30 @@ def getsourcefile(obj):
         return ret
 
 
-__all__ = ["getsourcelines", "getsourcefile"]
+class FunctionSourceInfo(BaseModel):
+    function_name: str
+    filepath: str
+    start_lineno: int
+    end_lineno: int
+
+    class Config:
+        frozen = True
+
+
+def get_source_info_and_src(func: Callable) -> tuple[FunctionSourceInfo, list[str]]:
+    file = getsourcefile(func)
+    name = func.__name__
+    src, start_lineno = getsourcelines(func)
+    end_lineno = start_lineno + len(src) - 1
+    return (
+        FunctionSourceInfo(
+            function_name=name,
+            filepath=file,
+            start_lineno=start_lineno,
+            end_lineno=end_lineno,
+        ),
+        src,
+    )
```

Comment:
what about:
```
func_info = FunctionSourceInfo(name, file, start_lineno, end_lineno)
return func_info, src
```

Reply:
I would add that I encourage using keyword when your define the variable where it is used, without storing it with an explicit name beforehand, i.e.:
```python
fortune = let_keep_dreaming(tutorial="".join((root_path, "my_rich.mp4")))
more_fortune = make_it_double(fortune)  # instead of make_it_double(fortune=fortune)
```

Reply:
> I'd prefer to keep the keywords on the whole. 

You can keep the keyword if you like. Still I would suggest to define `func_info = [...]` outside the return because formatting gets messy.